### PR TITLE
Implicit array casting

### LINKED ISSUE #8689: release-20.2: sql: allow implicit casting of string to more array types
PR: https://github.com/cockroachdb/cockroach/pull/54944

From release notes:
> A [string](../v20.2/string.html) literal like `'{X, Y, Z}'` is now automatically casted to an [`ARRAY`](../v20.2/array.html) when appropriate. Support is added for `UUID`, `DATE`, `BOOL`, `TIME`, `TIMETZ`, `TIMESTAMP`, `TIMESTAMPTZ`, `FLOAT`, `INET`, `VARBIT`, and `INTERVAL` arrays. Note that `INT` and `DECIMAL` arrays were already supported. [#54944][#54944] {% comment %}doc{% endcomment %}

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes #8689.

### CODE REVIEW

File: v20.2/data-types.md [hunk 1]
```diff
@@ -47,7 +47,7 @@ CockroachDB supports explicit type conversions using the following methods:
 
 - Other [built-in conversion functions](functions-and-operators.html) when the type is not a SQL type, for example `from_ip()`, `to_ip()` to convert IP addresses between `STRING` and `BYTES` values.
 
- CockroachDB also supports implicit casting from string literals to `INT` and `DECIMAL` [`ARRAY`](array.html)s, where appropriate. For an example, see [Implicit casting to `INT` and `DECIMAL` `ARRAY`s](array.html#implicit-casting-to-int-and-decimal-arrays).
+ CockroachDB also supports implicit casting from string literals to all data types except [`BYTES`](bytes.html), [`ENUM`](enum.html), [`JSONB`](jsonb.html), and [`SERIAL`](serial.html). For an example, see [Implicit casting to `ARRAY`s](array.html#implicit-casting-to-arrays).
```
Comment (rafiss) [inline v20.2/data-types.md:50]: should be "to arrays of all data types"

also, Box2D, Geometry, and Geography types are not supported

File: v20.2/array.md [hunk 2]
```diff
@@ -262,9 +262,9 @@ You can cast an array to a `STRING` value, for compatibility with PostgreSQL:
 (1 row)
 ~~~
 
-### Implicit casting to `INT` and `DECIMAL` `ARRAY`s
+### Implicit casting
 
- CockroachDB supports implicit casting from string literals to [`INT`](int.html) and [`DECIMAL`](decimal.html) `ARRAY`s, where appropriate.
+CockroachDB supports implicit casting from string literals to arrays of all data types except [`BYTES`](bytes.html), [`ENUM`](enum.html), [`JSONB`](jsonb.html), and [`SERIAL`](serial.html).
```
Comment (rafiss) [inline v20.2/array.md:267]: Box2D, Geometry, and Geography types are not supported too

File: v20.2/string.md [hunk 3]
```diff
@@ -113,7 +113,7 @@ The size of a `STRING` value is variable, but it's recommended to keep values un
 
 Type | Details
 -----|--------
-`ARRAY` | Requires supported [`ARRAY`](array.html) string format, e.g., `'{1,2,3}'`.<br>Limited to `ARRAY`s of type [`INT`](int.html) and [`DECIMAL`](decimal.html).
+`ARRAY` | Requires supported [`ARRAY`](array.html) string format, e.g., `'{1,2,3}'`.<br>Note that string literals can be implicitly cast to any supported `ARRAY` data type except [`BYTES`](bytes.html), [`ENUM`](enum.html), [`JSONB`](jsonb.html), and [`SERIAL`](serial.html).
```
Comment (rafiss) [inline v20.2/string.md:116]: same comment re: Box2D, Geometry, and Geography

**Review by rafiss — CHANGES_REQUESTED** [review]

**Review by ericharmeling — COMMENTED** [review]
TFTR @rafiss !

I updated the pages with those spatial types, and changed the formatting a bit. I don't think we currently document support for Box2D, but that could be in the works. @rmloveland 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8709#-:-MK1jLTN2CEBAQS5A_i6:b-yq3trf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[v20.2/array.md, line 267 at r1](https://reviewable.io/reviews/cockroachdb/docs/8709#-MK1JYUOmokUuUZl4de6-r1-267:-MK1hsZM97FqgnEiefEi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/d488effe1cb9d8f740bf598d1686b4570803b2dd/v20.2/array.md#L267)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

Box2D, Geometry, and Geography types are not supported too
</blockquote></details>

Done.
___
*[v20.2/data-types.md, line 50 at r1](https://reviewable.io/reviews/cockroachdb/docs/8709#-MK1JYUOmokUuUZl4de7-r1-50:-MK1i6Iy07KHOBtfGEik:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/d488effe1cb9d8f740bf598d1686b4570803b2dd/v20.2/data-types.md#L50)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should be "to arrays of all data types"

also, Box2D, Geometry, and Geography types are not supported
</blockquote></details>

Done.
___
*[v20.2/string.md, line 116 at r1](https://reviewable.io/reviews/cockroachdb/docs/8709#-MK1JYUOmokUuUZl4de8-r1-116:-MK1jKe84hJLsn6bqSFi:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/d488effe1cb9d8f740bf598d1686b4570803b2dd/v20.2/string.md#L116)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

same comment re: Box2D, Geometry, and Geography
</blockquote></details>

Done.

**Review by lnhsingh — COMMENTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/8709#-:-MK5V_T92A3nmLjLwFXT:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lnhsingh and @rafiss)

**Comment by rmloveland** [comment]
Eric Harmeling <notifications@github.com> writes:

> I updated the pages with those spatial types, and changed the
> formatting a bit. I don't think we currently document support for
> Box2D, but that could be in the works. @rmloveland

We don't have docs yet!  It's not on the "required shortlist" of things
for the release, but it Will Definitely Get Done Eventually (tm).

**Pull request merged**